File: src/main/java/org/taktik/icure/logic/impl/TarificationLogicImpl.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic.impl;


import java.util.Collections;
import java.util.List;

import com.google.common.base.Preconditions;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.taktik.icure.dao.TarificationDAO;
import org.taktik.icure.db.PaginatedList;
import org.taktik.icure.db.PaginationOffset;
import org.taktik.icure.entities.Tarification;
import org.taktik.icure.logic.TarificationLogic;

@Service
public class TarificationLogicImpl extends GenericLogicImpl<Tarification, TarificationDAO> implements TarificationLogic {
	private static final Logger logger = LoggerFactory.getLogger(TarificationLogicImpl.class);

	private TarificationDAO tarificationDAO;

	@Override
	public Tarification get(String id) {
		return tarificationDAO.get(id);
	}

	@Override
	public Tarification get(@NotNull String type, @NotNull String tarification, @NotNull String version) {
		return tarificationDAO.get(type + "|" + tarification + "|" + version);
	}

	@Override
	public List<Tarification> get(List<String> ids) {
		return tarificationDAO.getList(ids);
	}

	@Override
	public Tarification create(Tarification tarification) {
		Preconditions.checkNotNull(tarification.getCode(), "Tarification field is null.");
		Preconditions.checkNotNull(tarification.getType(), "Type field is null.");
		Preconditions.checkNotNull(tarification.getVersion(), "Version tarification field is null.");

		// assigning Tarification id type|tarification|version
		tarification.setId(tarification.getType() + "|" + tarification.getCode() + "|" + tarification.getVersion());

		return tarificationDAO.create(tarification);
	}

	@Override
	public Tarification modify(Tarification tarification) throws Exception {
		Tarification existingTarification = tarificationDAO.get(tarification.getId());

		Preconditions.checkState(existingTarification.getCode().equals(tarification.getCode()), "Modification failed. Tarification field is immutable.");
		Preconditions.checkState(existingTarification.getType().equals(tarification.getType()), "Modification failed. Type field is immutable.");
		Preconditions.checkState(existingTarification.getVersion().equals(tarification.getVersion()), "Modification failed. Version field is immutable.");

		updateEntities(Collections.singleton(tarification));

		return this.get(tarification.getId());
	}


	@Override
	public List<Tarification> findTarificationsBy(String type, String tarification, String version) {
		return tarificationDAO.findTarifications(type, tarification, version);
	}

	@Override
	public List<Tarification> findTarificationsBy(String region, String type, String tarification, String version) {
		return tarificationDAO.findTarifications(region, type, tarification, version);
	}

	@Override
	public PaginatedList<Tarification> findTarificationsBy(String region, String type, String tarification, String version, PaginationOffset paginationOffset) {
		return tarificationDAO.findTarifications(region, type, tarification, version, paginationOffset);
	}

	@Override
	public PaginatedList<Tarification> findTarificationsByLabel(String region, String language, String label, PaginationOffset paginationOffset) {
		return tarificationDAO.findTarificationsByLabel(region, language, label, paginationOffset);
	}

	@Override
	public PaginatedList<Tarification> findTarificationsByLabel(String region, String language, String type, String label, PaginationOffset paginationOffset) {
		return tarificationDAO.findTarificationsByLabel(region, language, type, label, paginationOffset);
	}

	@Override
	public Tarification getOrCreateTarification(String type, String tarification) {
		List<Tarification> tarifications = findTarificationsBy(type, tarification, null);

		if (tarifications.size()>0) {
			return tarifications.stream().sorted((a,b)->b.getVersion().compareTo(a.getVersion())).findFirst().get();
		}

		return this.create(new Tarification(type,tarification,"1.0"));
	}

	@Override
	protected TarificationDAO getGenericDAO() {
		return tarificationDAO;
	}

	@Autowired
	public void setTarificationDAO(TarificationDAO tarificationDAO) {
		this.tarificationDAO = tarificationDAO;
	}
}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/TarificationLogicImpl.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.asynclogic.impl

import com.google.common.base.Preconditions
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.toList
import org.springframework.stereotype.Service
import org.taktik.couchdb.ViewQueryResultEvent
import org.taktik.icure.asyncdao.TarificationDAO
import org.taktik.icure.asynclogic.AsyncSessionLogic
import org.taktik.icure.asynclogic.TarificationLogic
import org.taktik.icure.db.PaginationOffset
import org.taktik.icure.entities.Tarification

@Service
class TarificationLogicImpl(private val tarificationDAO: TarificationDAO, private val sessionLogic: AsyncSessionLogic) : GenericLogicImpl<Tarification, TarificationDAO>(sessionLogic), TarificationLogic {

    override suspend fun getTarification(id: String): Tarification? {
        return tarificationDAO.get(id)
    }

    override suspend fun getTarification(type: String, tarification: String, version: String): Tarification? {
        return tarificationDAO.get("$type|$tarification|$version")
    }

    override fun getTarifications(ids: List<String>): Flow<Tarification> = flow {
        emitAll(tarificationDAO.getEntities(ids))
    }

    override suspend fun createTarification(tarification: Tarification) = fix(tarification) { tarification ->
        tarification.code ?: error("Code field is null")
        tarification.type ?: error("Type field is null")
        tarification.version ?: error("Version field is null")

        // assigning Tarification id type|tarification|version
        tarificationDAO.create(tarification.copy(id = tarification.type + "|" + tarification.code + "|" + tarification.version))
    }

    override suspend fun modifyTarification(tarification: Tarification) = fix(tarification) { tarification ->
        val existingTarification = tarification.id?.let { tarificationDAO.get(it) }
        Preconditions.checkState(existingTarification?.code == tarification.code, "Modification failed. Tarification field is immutable.")
        Preconditions.checkState(existingTarification?.type == tarification.type, "Modification failed. Type field is immutable.")
        Preconditions.checkState(existingTarification?.version == tarification.version, "Modification failed. Version field is immutable.")
        modifyEntities(setOf(tarification)).firstOrNull()
    }

    override fun findTarificationsBy(type: String?, tarification: String?, version: String?): Flow<Tarification> = flow {
        emitAll(tarificationDAO.listTarificationsBy(type, tarification, version))
    }

    override fun findTarificationsBy(region: String?, type: String?, tarification: String?, version: String?): Flow<Tarification> = flow {
        emitAll(tarificationDAO.listTarificationsBy(region, type, tarification, version))
    }

    override fun findTarificationsBy(
        region: String?,
        type: String?,
        tarification: String?,
        version: String?,
        paginationOffset: PaginationOffset<List<String?>>
    ): Flow<ViewQueryResultEvent> = flow {
        emitAll(tarificationDAO.findTarificationsBy(region, type, tarification, version, paginationOffset))
    }

    override fun findTarificationsByLabel(region: String?, language: String?, label: String?, paginationOffset: PaginationOffset<List<String?>>): Flow<ViewQueryResultEvent> = flow {
        emitAll(tarificationDAO.findTarificationsByLabel(region, language, label, paginationOffset))
    }

    override fun findTarificationsByLabel(region: String?, language: String?, type: String?, label: String?, paginationOffset: PaginationOffset<List<String?>>): Flow<ViewQueryResultEvent> = flow {
        emitAll(tarificationDAO.findTarificationsByLabel(region, language, type, label, paginationOffset))
    }

    override suspend fun getOrCreateTarification(type: String, tarification: String): Tarification? {
        val listTarifications = findTarificationsBy(type, tarification, null).toList()
        return listTarifications.takeIf { it.isNotEmpty() }?.let { it.sortedWith(Comparator { a: Tarification, b: Tarification -> b.version!!.compareTo(a.version!!) }) }?.first()
                ?: createTarification(Tarification.from(type, tarification, "1.0"))
    }

    override fun getGenericDAO(): TarificationDAO {
        return tarificationDAO
    }

}
